### PR TITLE
PS-7748: Implement buffered error logging ability

### PR DESCRIPTION
**Pull request by dutow**
This commits adds two new variables, buffered_error_log_filename and
buffered_error_log_size.

When buffered_error_log_size is specified and non zero, a buffer is
allocated with the specified size, and certain (debug) log messages are
written there.

When this buffer gets full, the server shuts down, or crashes, the
buffer is written to buffered_error_log_filename.

This is useful in situations where debug logs create lots of output and
slow down normal server operation, but are helpful for figuring out
issues, as writing to a memory buffer has less overhead.

### CODE REVIEW

**Comment by george-lorch** [comment]
PXC-3574 is "Port wsrep_Mode variable and Functionality from Galera 8.0.22-26.5"  Do you mean PS-7748? :)

**Comment by dutow** [comment]
Ops, yes, I mixed up two jira tabs when writing the commit message. Fixed.

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: @dutow the option names are not very clear. Can we use something like memory_log_filename or memory_log_filesize
or mem_log_filename, or mem_log_filesize or shared_mem_log_filename?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
sorry for the delay, will start the review tomorrow

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: I'm not sure about changing size to filesize, as this is primarily a memory buffer. We only write it out into a file if the server crashes.

As for changing blackbox to memory_log or mem_log, we can do that, I only used blackbox because that's the name mariadb uses.

File: mysys/blackbox.h
Comment (satya-bodapati) [inline mysys/blackbox.h]: dumped*

File: mysys/blackbox.h
Comment (satya-bodapati) [inline mysys/blackbox.h]: any reason why std::vector<char> is preferred to std::string?

File: mysys/blackbox.h
Comment (satya-bodapati) [inline mysys/blackbox.h]: whats the TODO left here?

File: mysys/blackbox.h
Comment (satya-bodapati) [inline mysys/blackbox.h]: why not std::lock_guard?

File: sql/sys_vars.cc
Comment (satya-bodapati) [reply sql/sys_vars.cc]: Did we take the implementation from MariaDB?

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [inline sql/server_component/log_sink_trad.cc]: remove this extra comment line

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
@dutow CircleCI (clang-format) test failed?
Also please provide jenkins link

**Comment by satya-bodapati** [comment]
@dutow Regarding the performance impact, can we get some numbers?

I was reading the Logger functionality code and found that we do not strictly write every line to stderr and flush. There is a buffering logic and based on some interval duration, the buffer is flushed to disk.

Is it possible to adjust that time based logic to improve performance? Do we need this feature at all? :smile:

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: No. This is only in mariadb enterprise, and they used a shared memory buffer.

Comment (satya-bodapati) [reply sql/sys_vars.cc]: Great! then we are free to choose whatever we want :+1:

**Comment by dutow** [comment]
@satya-bodapati I think flushing is only part of the issue, the amount of log is also there. PXC with wsrep_debug>0, especially with 4 will result in huge log files.

wsrep_on=1 results in a 25% increased execution time, 12% with this feature. If I only modify the original code by removing the actual file writing calls (logging calls are done, but nothing is written to he memory/disk) that result in 10-11%, which means that this blackbox logging feature only adds a 1 percent overhead over that. And most of that 10% is spent in the string formatting code, we can't do anything about that.

With wsrep=4 these numbers are 70%+ without, 39% with this code.

I got these numbers with a release build, and writing logs to a PCIE4 NVME SSD. If I store the logs on a slower SSD, it is much worse.

**Comment by dutow** [comment]
Also, generally you only need the content of these additional logs when a crash occurs. During normal operation these just add additional noise to the normal server logs.

**Comment by satya-bodapati** [comment]
> @satya-bodapati I think flushing is only part of the issue, the amount of log is also there. PXC with wsrep_debug>0, especially with 4 will result in huge log files.
> 
> wsrep_on=1 results in a 25% increased execution time, 12% with this feature. If I only modify the original code by removing the actual file writing calls (logging calls are done, but nothing is written to he memory/disk) that result in 10-11%, which means that this blackbox logging feature only adds a 1 percent overhead over that. And most of that 10% is spent in the string formatting code, we can't do anything about that.
> 
> With wsrep=4 these numbers are 70%+ without, 39% with this code.
> 
> I got these numbers with a release build, and writing logs to a PCIE4 NVME SSD. If I store the logs on a slower SSD, it is much worse.

Ok, if there is a way to measure performance, then it is really good. 
What about messages in ring buffer on recycle? Do we discard them?

Also, can you please verify how many messages are in buffer for each flush. I think this buffering isn't happening in all cases? :thinking: 
in log_sink_trad(), please check about this part

```
   341               if (out_types & LOG_ITEM_LOG_BUFFERED) {                                                                                                       │
│   342                 // abort if we've already printed this (to trad log only!)                                                                                   │
│   343                 if (microtime <= log_sink_trad_last) {                                                                                                       │
│   344                   out_fields = LOG_SERVICE_NOTHING_DONE;                                                                                                     │
│   345                   goto done;  
```

**Comment by dutow** [comment]
@satya-bodapati Looks like that buffering is only there for early server startup. As soon as the logging system is properly initialized, it is no longer used.

For the ring buffer: we only keep the messages that fit into the buffer, and discard earlier ones. The idea is that users can set a buffer size which will most likely contain enough messages when an error occurs.

**Comment by satya-bodapati** [comment]
> @satya-bodapati Looks like that buffering is only there for early server startup. As soon as the logging system is properly initialized, it is no longer used.
> 
> For the ring buffer: we only keep the messages that fit into the buffer, and discard earlier ones. The idea is that users can set a buffer size that will most likely contain enough messages when an error occurs.

@dutow, is it possible to implement a similar buffer logic and write that buffer after it is filled? 

Regarding the current ring buffer logic, I feel it is very hard for users to estimate the buffer size required. For example, how much should be the buffer size if I want to log messages that are generated in a day?

Hmm, can we write ring buffer before discarding it? (same request as above, but I guess this change will be in your new code?)

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: I've changed the name to buffered_log. With the ringbuffer / discarding behavior gone, I think that's a good description.

File: mysys/blackbox.h
Comment (dutow) [reply mysys/blackbox.h]: fixed.

File: mysys/blackbox.h
Comment (dutow) [reply mysys/blackbox.h]: Only that the size of the vector is more explicit than the capacity of a string, but that's not that important, I changed it to string.

File: mysys/blackbox.h
Comment (dutow) [reply mysys/blackbox.h]: Nothing, removed.

File: mysys/blackbox.h
Comment (dutow) [reply mysys/blackbox.h]: Changed to lock_guard.

File: sql/server_component/log_sink_trad.cc
Comment (dutow) [reply sql/server_component/log_sink_trad.cc]: removed.

File: mysys/buffered_log.h
Comment (satya-bodapati) [inline mysys/buffered_log.h]: is there a way to assert mutex is taken?

File: mysys/buffered_log.h
Comment (satya-bodapati) [inline mysys/buffered_log.h]: better to follow same style w.r.t to braces for both if condidtions

File: mysys/buffered_log.h
Comment (satya-bodapati) [inline mysys/buffered_log.h]: Should we keep closing the file descriptor after each buffer flush?

File: mysys/buffered_log.h
Comment (satya-bodapati) [inline mysys/buffered_log.h]: what will happen if we dont do this? write \0 to file?

File: mysys/buffered_log.h
Comment (satya-bodapati) [inline mysys/buffered_log.h]: performance

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: I still see it

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [inline sql/server_component/log_sink_trad.cc]: what is the use of this variable? Can be removed?

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: I think it we should mention error_log instead of just log

buffer_error_log_filename

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
@dutow Looks really good. Few suggestions

Now that we have decided on new names, Please fix
1. Github description
2. Remove blackbox everywhere
3. Commit message

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: buffered_error_log_size

Also please change all variable name to indicate error or err

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: log_only_to_buffered_log

File: mysys/buffered_log.h
Comment (dutow) [reply mysys/buffered_log.h]: No, this only affects the length calculation, it's not that critical. This len is only used for the buffer full check below it. It was more important with the vector version, where we did write the \0 to the file otherwise.

In the current code I could remove it, and then sometimes we might flush the buffer one message earlier than we should, that doesn't really matter.

Comment (satya-bodapati) [reply mysys/buffered_log.h]: Ok, will leave it to you :) if you want to retain, please add comment

File: mysys/buffered_log.h
Comment (satya-bodapati) [reply mysys/buffered_log.h]: Just check. Not very important.

File: sql/server_component/log_sink_trad.cc
Comment (dutow) [reply sql/server_component/log_sink_trad.cc]: No, these both have real usecases.

log_to_buffered_log tells us if we should write something to the buffered log.
log_only_to_buffered_log tells us that we shouldn't write that message to the normal error log, only to the buffered log.

We definitely need log_only_to_buffered_log. 

Log_to_buffered_log is a different question. For performance testing, I've written all log messages to the buffered error log too, but we might have use cases where we only want to write some of the log messages there. The mariadb implementation is like that (based on the documentation), they only write wsrep messages to the blackbox, that's why both variables are here.

File: mysys/buffered_log.h
Comment (dutow) [reply mysys/buffered_log.h]: I don't think there's any without a custom wrapper class around std::mutex, or adding an std::unique_lock as a data member. std::mutex doesn't have this functionality. Even try_lock + unlock doesn't work, because calling try_lock if the current thread already owns the mutex is UB.

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: I mean, it will be compiler optimized anyway. I see the variable used
bool x  = false
if (!false || ..)

Comment (dutow) [reply sql/server_component/log_sink_trad.cc]: Because in the current code, there's nothing using this logging other than that DBUG_EXECUTE_IF code (and that only checks that we write to the buffered log correctly).

PXC will set log_only_to_buffered_log true if we are writing a wsrep/galera log message above the warning level.

File: mysys/buffered_log.h
Comment (satya-bodapati) [reply mysys/buffered_log.h]: ok, ignore this

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: So if I understand correctly, there are two functionalities we want to achieve?
1. Bufffered logging (log to buffer and write to disk when buffer is full)
2. Buffer-only Logging (write only to buffer, when buffer is full, we discard/overwrite. Buffer is written to disk only at crash)?

Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: And regarding 1. the buffered logging, can't we use same error log file ? ie. write buffered messages to default mysqld error log.

Comment (dutow) [reply sql/server_component/log_sink_trad.cc]: No, we only want 2. Changing the basic error logging isn't a goal of this PR, as that doesn't produce enough output to be worth it.

But we want to be able to do "2" in two ways:

* write normal messages to the classic error log. Write both normal and high output debug messages to the buffered error log
* write normal messages to the classic error log. Write high output debug messages to the buffered error log

File: mysys/buffered_log.h
Comment (dutow) [reply mysys/buffered_log.h]: Good question. I implemented this way because (depending on the buffer size) writes aren't that frequent. With this the user should also be free to move / rotate / etc the logfile,as we always reopen it before writing to it.

File: mysys/buffered_log.h
Comment (dutow) [reply mysys/buffered_log.h]: I removed it, it shouldn't cause any noticable difference.

File: mysys/buffered_log.h
Comment (dutow) [reply mysys/buffered_log.h]: fixed

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: If not for log rotation, I would love to see a simpler logic. Just use buffered functionality for all messages and write buffer to mysqld error log. 

Maybe we should keep the log rotation problem separate and fix that separately. WIth that my proposal would be

1. No separate error log file (no option for this)
2. If buferred_error_log_buffer_size is not 0, use buffering and write the buffer to mysqld error log file.
3. If buferred_error_log_buffer_size is 0, normal logging (as if this feature is disabled)

File: mysys/buffered_error_log.h
Comment (satya-bodapati) [inline mysys/buffered_error_log.h]: upstream style is Buffered_error_logger

File: mysys/buffered_error_log.h
Comment (satya-bodapati) [inline mysys/buffered_error_log.h]: should we move implementation to .cc file?

File: mysys/buffered_error_log.h
Comment (satya-bodapati) [inline mysys/buffered_error_log.h]: mysql-secure-file-priv check missing?

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: Better to keep these separate and all variables in same codeblock

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: Also put them in alphabetical sort position? :thinking:

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [inline sql/server_component/log_sink_trad.cc:395]: Should we add the control flag elsewhere?
LIke this?: LogEvent().module("blah").buffered(true)

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: @dutow what do u think? u want to keep buffered log file as separate?

File: sql/server_component/log_sink_trad.cc
Comment (dutow) [reply sql/server_component/log_sink_trad.cc:395]: I'm not sure about this, as the idea is to log everything printed by a submodule or loglevel to this log. That condition is easier to implement directly in log_sink_trad instead of adding it to all the macros/places log messages are generated.

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: That isn't an alphabetical list, entries before this are also in a random order.

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: Fixed.

File: mysys/buffered_error_log.h
Comment (dutow) [reply mysys/buffered_error_log.h]: It isn't used in any of the log variables. They use a different log file name check functions, I've added that.

File: mysys/buffered_error_log.h
Comment (dutow) [reply mysys/buffered_error_log.h]: Done.

File: mysys/buffered_error_log.h
Comment (dutow) [reply mysys/buffered_error_log.h]: Changed.

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: @dutow please let me know your thoughts on this (avoiding separate error log file) :)

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: I think we should not allow arbirtray sizes. what if we set just 100 bytes? and a single error message needs 200 bytes?
May be multiples of 16K or 1MB? or start value for which we saw decent improvement?

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc:395]: ok, lets leave it as is

File: mysys/buffered_error_log.cc
Comment (satya-bodapati) [inline mysys/buffered_error_log.cc:32]: when do we come log and the buffer capacity is null or 0?

File: sql/sys_vars.cc
Comment (satya-bodapati) [reply sql/sys_vars.cc]: Also, if the buffer is 1MB, the buffer is filled only until 512K, we will never see the remaining 512k until shutdown or crash?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
few concerns on the buffer size length possibilities
Also please check my question on alternate possibility (avoiding separate error log file)

File: sql/server_component/log_sink_trad.cc
Comment (dutow) [reply sql/server_component/log_sink_trad.cc]: Yes, I'd like to keep it separe for several reasons:

1. buffered logging is delayed. Users might want to reveive normal log messages without significant delays.
2. buffered logging is intended for debug logs (high throughput, huge logfiles). This is only required if something bad happens, and we want to investigate how that happened, while the normal logs could be useful for normal server operation. Also, because of the larger size, keeping it separate means users can use different rotation/size limits (e.g. keep normal logs for a few months, but keep buffered logs only for the last day, or something like that)

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: Yes, we only write on buffer full and on exit.

File: mysys/buffered_error_log.cc
Comment (dutow) [reply mysys/buffered_error_log.cc:32]: If `log_to_buffered_error_log` is true, but the user didn't configure buffered logging.

File: sql/sys_vars.cc
Comment (satya-bodapati) [reply sql/sys_vars.cc]: @dutow can you please tell what happens when buffer size is 100 bytes and a single error message is 200 bytes?

File: sql/server_component/log_sink_trad.cc
Comment (satya-bodapati) [reply sql/server_component/log_sink_trad.cc]: ok @dutow I think the answers are convincing enough to keep it as is

File: sql/sys_vars.cc
Comment (dutow) [reply sql/sys_vars.cc]: Since we are using an std::string as buffer and operator+= for adding the new log line to the buffer, it will automatically extend the buffer.

**Review by satya-bodapati — APPROVED** [review]
LGTM! Please ensure clean jenkins run before merging

File: mysys/buffered_error_log.h
Comment (satya-bodapati) [reply mysys/buffered_error_log.h]: @dutow I think, it is important to add file-priv check. THis is because, other log-file variables are startup and read-only. Since this is a dynamic option, we have to do privilege check. Apart from that, we have to ensure that the file pointed doesn't overwrite an existing file? What if the user points to mysql.ibd? or any other file inside datadir?

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
@dutow please check my comment on privilege check and other vulnerabilities.

File: mysys/buffered_error_log.h
Comment (satya-bodapati) [reply mysys/buffered_error_log.h]: @dutow Please check if the above validations are done for https://dev.mysql.com/doc/refman/5.7/en/server-system-variables.html#sysvar_general_log_file?

File: sql/sys_vars.cc
Comment (satya-bodapati) [inline sql/sys_vars.cc]: @dutow Change looks good. Please add some negative tests with buffered_error_log and the sys_vars.test

**Review by satya-bodapati — CHANGES_REQUESTED** [review]

**Comment by dutow** [comment]
@satya-bodapati I rebased this based on the now merged log restricttion patch.

File: mysql-test/t/buffered_error_log.test
Comment (satya-bodapati) [inline mysql-test/t/buffered_error_log.test:1]: @dutow why did you suppress this? Does it mean there was assertion failure and core file generated?

File: mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result:9]: when the buffered file name is empty, what is the filename chosen?

File: mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result
Comment (satya-bodapati) [inline mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result]: if you use GLOBAL.buffer_err_log_size=OFF, will it accept?

File: mysql-test/t/buffered_error_log.test
Comment (dutow) [reply mysql-test/t/buffered_error_log.test:1]: The test intentionally crashes the server. (to test that we do write the log)

File: mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result
Comment (dutow) [reply mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result:9]: Nothing, then it's off.

File: mysql-test/t/buffered_error_log.test
Comment (satya-bodapati) [reply mysql-test/t/buffered_error_log.test:1]: simple, kill_mysqld.inc is not enough?

File: mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result
Comment (satya-bodapati) [reply mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result:9]: ok, makes sense. Can you show the output of filename instead of error_log_size?

File: mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result
Comment (dutow) [reply mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result]: No, it fails with incorrect argument type. I'll change this to 0 too.

File: mysql-test/t/buffered_error_log.test
Comment (dutow) [reply mysql-test/t/buffered_error_log.test:1]: It's not, it doesn't suppress anything. You can see this suppression in many tests that use it.

File: mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result
Comment (dutow) [reply mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result:9]: I don't understand the question, the select below shows the filename? This file doesn't mention size at all.

File: mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result
Comment (dutow) [reply mysql-test/suite/sys_vars/r/buffered_error_log_size_basic.result]: Fixed.

File: mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result
Comment (satya-bodapati) [reply mysql-test/suite/sys_vars/r/buffered_error_log_filename_basic.result:9]: sorry, I didn't see that you are showing filename.

File: mysql-test/t/buffered_error_log.test
Comment (satya-bodapati) [reply mysql-test/t/buffered_error_log.test:1]: I mean, instead of send_kill_to_mysqld.inc, why not use kill_mysqld.inc which sends signal 9?

Comment (satya-bodapati) [reply mysql-test/t/buffered_error_log.test:1]: ok, we can want to check if error log is saved to disk on core file generation/assert failures. Got it

**Review by satya-bodapati — APPROVED** [review]

**Comment by kamil-holubicki** [comment]
My understanding is that it works in the way:
1. Log only to ram-buffered log
2. Log only to the disk log
3. Log to 1 AND 2
No filtering, logs splitting based on log severity, and so on.

Just a guess:
In this comment https://github.com/percona/percona-server/pull/4358#issuecomment-891560720 you showed benchmark analysis. Do you have similar results for locating the standard error log on /dev/shm? Do we still have the performance problem there? If not, maybe it would be better to implement the functionality for circular logging for the standard logging system?


In other comment I read:
```wsrep_on=1 results in a 25% increased execution time, 12% with this feature. If I only modify the original code by removing the actual file writing calls (logging calls are done, but nothing is written to he memory/disk) that result in 10-11%```
Do I understand correctly that original logging, if disk write is removed is as fast as the new feature? If yes, I would check the behavior with /dev/shm and consider adding circular logging feature for standard logging system.

**Comment by dutow** [comment]
@kamil-holubicki I did consider that, but that would mean delaying normal log messages, while users might expect them to appear in time as usual.

This solution is a compromise: it keeps the existing log functioning as is, while adding a second sink for bulk output that we don't expect to be readable immediately.

**Pull request merged**
